Geometry Centre Optimisation

RobertRosca opened the pull request:
Re-submitted pr with different branch. Previous branch has a pure pyFAI implementation I'd like to archive (where extra-geom is not used to assemble the image and the pixel corner positions are passed to pyFAI directly), but that ended up being slower than this method.

Adds in a way to perform centre optimisation, which is crucial for analysis involving azimuthal integration. The tutorial notebook explains it better than I can here: https://github.com/RobertRosca/geoAssembler/blob/optimiser/docs/geoOptimiser-tutorial.ipynb

Main things added are:

`Integrator` class:
 - Takes in our geometry (+ sample distance in mm) object and wraps the pyFAI functions
 - Makes it a bit cleaner/easier to set up and run the integrator
 - An identical class exists in the update of the `extra-data-interactive` widget, however as that is only on GitLab I have defined it here again. Would it be worth having something like `extra-algorithms` where we can store these small-ish snippets? Could contain things like this integrator, and known ways to mask detectors, etc...

`CentreOptimiser` class:
 - Takes in our geometry and a detector image frame (+ sample dist)
 - Creates and stores an integrator object
 - Contains the loss function, function which creates and fills in the grids, and an optimise function which runs a coarse grid search followed by progressively finer searches

And finally the tutorial notebook mentioned at the start.

Marked as WIP as there are a some things left to do:

 - [x] Plotting functions - add in line to show that optimised version is straighter
 - [x] Convenient way to extract/save the corrected geometry
 - [x] Maybe some warnings for users trying to use multiprocessing on login nodes?
 - [ ] Come up with tests for the scipy optimiser - probably only need to see that the loss function is working correctly

takluyver commented on geoAssembler/tests/test_centre_optimiser.py:
```diff
@@ -0,0 +1,68 @@
+import ..optimiser as geoOptimiser
```
The test suit is complaining that this isn't valid - you can do `from .. import optimiser`

takluyver commented:
Looking at the notebook first - this is a pretty nice introduction.

- Let's avoid the name geoOptimiser (too close to the existing tool geoptimiser). centreOptimiser? recentrer? onering (torulethemall)?
- The intro should say that this works with sharp diffraction rings, so people can quickly tell if they have suitable data to use it.
- I'll look at the `avg_frame` function soon, but I'd prefer to make the example work through from loading data and assembling a frame with EXtra-geom, even if this requires a bit of extra code in the notebook or makes the pictures slightly less neat. The handy function with AGIPD-specific adjustments seems like a distraction from the key message.
- If the optimiser takes a geometry object, maybe it would be neater for it to take a 3D array and do the assembly as part of its setup.
- `missaligned` -> `misaligned`
- Is it possible to briefly describe what the optimiser does in the notebook? It's making the highest peak in the azimuthal integration as high as possible, right? Or is it more complex now?
- Corrected geometry: is this available from `res - something like `res.new_geom` - rather than copying and pasting it back into code?
- "...which typically do not have to be looked at." - a bit tortured! "...which you typically don't need to look at"?
- In the second 'unrolled' image, is it worth drawing a vertical line to show how neatly the rings now unroll? It could be offset a few pixels so it doesn't obscure the lines, or use a dotted line. Maybe even draw it under the image so you see it in the gaps.
- Let's add a header (maybe "Troubleshooting") where the discussion of examining the optimisation process starts. It's definitely good to include this, but it's also nice to have a clear marker where the standard workflow ends.

takluyver commented on geoAssembler/optimiser/centre.py:
```diff
@@ -0,0 +1,180 @@
+        self.integrate2d = self.integrator.integrate2d
+
+        #  Slightly dodgy way to pull the quadrant corner positions out of geom
+        #  TODO: Suggest adding this in to extra-geom?
```
Yes, I must remember to figure out how to add a notion of quadrant positions in EXtra-geom.

In the meantime, can you use `/ geom.pixel_size` instead of `* geom._get_plot_scale_factor('px')`? That looks a little bit less hacky.

takluyver commented on geoAssembler/optimiser/centre.py:
```diff
@@ -0,0 +1,180 @@
+        so that it can be minimised.
+
+        Note: the first and last 100 radial bins are excluded, as these can
+        lead to artefacts which cause the optimisation to fail.
```
Are there guaranteed to always be > 200 bins? Is the failure comprehensible if there aren't?

RobertRosca replied:
Yep, the number of bins comes from this kind of silly bit:  `self.radius = ((self.size[0]/2)**2 + (self.size[1]/2)**2)**(1/2)`, so for assembled AGIPD images it should always be at least 700 ish I expect.

takluyver commented on geoAssembler/optimiser/centre.py:
```diff
@@ -0,0 +1,180 @@
+                centre_offset,
+                verbose=verbose,
+                pool=pool
+            )
```
I'm still in two minds about the 'zoom in and enhance' custom optimiser here. It does mean we can understand exactly what the code is doing, and easily make plots to inspect the intermediate steps. And it seems to work, which is never something to sneeze at!

On the other hand, numerical optimisation is a well established thing, and there are many battle-tested algorithms for doing it. [scipy.optimize](https://docs.scipy.org/doc/scipy/reference/optimize.html) has several, and there's also [scikit-optimize](https://scikit-optimize.github.io/stable/), which includes [visualisation tools](https://scikit-optimize.github.io/stable/modules/plots.html) for inspecting what's happening. Doing something smarter than grid search might mean it's easy to extend this to find quadrant positions (i.e. fitting 8 variables).

Can I ask you to investigate what it would look like to use an off-the-shelf optimiser? Is it easy to fit it into the code, and what do we lose/gain in inspecting the results? I might start with scikit-optimize, which looks more approachable than scipy. If you don't have time for this, I don't mind landing the feature with the custom optimiser, but I'm keeping my fingers crossed that it would be an easy win.

RobertRosca replied:
Yeah I looked into this a bit and just wrote my own as it was easier than figuring out how to use the existing libraries.

Far as I remember when I originally did this, none of the scipy optimisers work for this as they're designed for continuous/floating value optimisation, and the steps/gradients for that are hard coded into the optimisers. scikit optimize does have a way to define an [integer search space](https://scikit-optimize.github.io/stable/modules/generated/skopt.space.space.Integer.html), but that does (far as I could tell from the docs) the same thing mine does, then I [read the the "docs" ](https://scikit-optimize.github.io/stable/modules/space.html#space) and given the info there is seemed easier to make my own.

takluyver replied:
> one of the scipy optimisers work for this as they're designed for continuous/floating value optimisation

I'm a bit puzzled - isn't the centre location a continuous variable? I'd expect pyFAI to be able to use greater than pixel accuracy for this. Not that we necessarily need that precision, given that we're 'snapping' pixels to a single grid anyway, but it should mean we're optimising in a continuous space.

RobertRosca replied:
Ah yes I didn't explain that well, I still wanted to restrict it to integer space at the start and to have it only look at sub-pixel regions later, since there I imagined an optimiser might get stuck looking at the sub-pixel positions too early. Now that I actually looked into it it's actually very easy to implement, turns out I was looking in the wrong part of the docs :P

Summary of my adventures would be that all but one of the optimisation methods did not work well or at all for this problem. In the end the differential evolution method of global optimisation from scipy finished in a vaguely sensible amount of time, actually gave good results, **and** has multiprocessing via workers built in, so I'll swap over to using that. 

Thanks for prompting me to look into this more, I originally tried only the basic optimiser, saw it didn't work, and wrote them off. Turns out the penultimate method in the docs is the one that works lol.

--------------------------------

This is a longer summary, mostly for myself for future reference, so feel free to skip it.

I tried `gp_minimize` but everything I tried ended up being both wwaayy slower, and the results were worse as well. The Bayesian optimisation it uses really likes getting stuck at the edges of the bounds, even though the minimum isn't near the edges, I really don't get why it's behaving like that, spent a while trying to fix it and gave up. It also has the (to me) very odd behaviour of repeatedly testing the same argument, so it doesn't seem to cache results or be aware of what regions it has already gone over.

Also tried a few of the scipy optimize minimisation methods, went through most of the unconstrained and constrained multivariate function minimizers, and none of them worked very well out of the box either.

That was... vaguely predictable I guess, the function that's being minimised isn't the best and jumps around a fair bit, so I moved on to trying to use global optimisers.

SHGO was fast, but didn't find the minimum. Dual annealing was slow to the point where I gave up running it, and there is no simple way to parallelise it; same with basin hopping.

I also found that scipy optimize has a 'brute' method, which does exactly what I do with my grid search method.

Mercifully in the end the differential evolution method  finished in a vaguely sensible amount of time, actually gave good results, **and** has multiprocessing via workers built in, so I'll swap over to using that.

RobertRosca commented on geoAssembler/optimiser/utility.py:
```diff
@@ -0,0 +1,131 @@
+
+    Creating the integrator requires a geometry, as well as the distance
+    from the sample to the detector in mm. If unknown, 200mm seems to be
+    a good 'normal' guess for AGIPD.
```
Yeah it kind of matters, it's a required argument so you need to put something in there anyway, and changing the sample distance changes what radius the rings appear at in the unrolled image.

If the value is very large/small all of the rings get squished together the centre/edge of the image, which makes sense: if your sample is very far away then the detector would show that the rings have a larger radius, if it's very close they'd have a smaller radius, and pyFAI corrects for this.

Ideally the actual distance would be used, but if not then 200mm seems to do decently well for all the powder diffraction data I've looked at.

takluyver replied:
OK, fair enough.

RobertRosca commented on geoAssembler/optimiser/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from . import (utility, centre)
```
Removed the unnecessary import

RobertRosca commented on geoAssembler/optimiser/utility.py:
```diff
@@ -0,0 +1,131 @@
+    optionally can provide a slice as the pulse pattern if required.
+
+    Returns a 2d image, and the centre of the image as given by the geometry.
+    """
```
I agree :P how do you think it should be implemented? I guess just as a property that returns ones for the asic edges, something like `geom.asic_edges`, so that users could do `assembled_image[geom.asic_edges] = np.nan`, or deal with them however they want?

As for removing bad values, makes sense, I'll take it out of the code and move it to the example with an explanation. It was in the code as if you don't try to deal with the high error values somehow then even a single pixel at a ridiculously high value breaks the optimisation, so there wasn't a real scenario where you wouldn't want to mask high values off, but it should be more explicitly stated that this is being done.

takluyver replied:
I think I'd do it as a standalone function `agipd_asic_edges()`, returning a (512, 128) array distinguishing the edge pixels from the normal pixels. It's the same for every module, so you can apply that to a stacked array using numpy broadcasting, and then do the image assembly afterwards.

RobertRosca commented on geoAssembler/optimiser/utility.py:
```diff
@@ -0,0 +1,131 @@
+                self.sample_dist_mm,
+                centre_offset[0] + self.centre[0],
+                centre_offset[1] + self.centre[1]
+            )
```
Made it stateless, and changed from set_FIT2D to set_pyFAI, however this has (again) confused me about which index is x and y in the centre and in the offset.

Moving from `set_FIT2D` to `set_pyFAI` meant I had to change the order of the centre and centre offsets, now I'm a bit confused about the correct order of x and y, but the tests pass and the notebook produces the same results even though there's probably a mistake somewhere in the code.

Think two of my mistakes cancel out, I'll look into this a bit more later :wink:

takluyver replied:
Yep, coordinate order is a regular source of headaches. It's all because images tend to be stored row by row, so you index pixels (y, x) in languages that follow C rather than Fortran.

RobertRosca commented on geoAssembler/optimiser/centre.py:
```diff
@@ -0,0 +1,180 @@
+    straight lines in polar coordinates indicates an accurate centre.
+    """
+    def __init__(self, geom: DetectorGeometryBase,
+                 frame: np.ndarray, sample_dist_mm: Union[int, float],
```
Hmmm, going off of [good old Stackoverflow](https://stackoverflow.com/questions/50928592/mypy-type-hint-unionfloat-int-is-there-a-number-type), the 'correct' thing to do is to use `float` as that covers int as well, this is very counterintuitive to me as the hint you're given is to put a float in, when really I want it to explicitly show that it accepts an int or a float

As for `Real`, there is a [bug which doesn't look like it'll get fixed any time soon](https://github.com/python/mypy/issues/3186) which means that using `Real` to cover any kind of real numbers doesn't actually work.

So it looks like using `Union[int, float]` is still the best option for now.

RobertRosca commented:
> Let's avoid the name geoOptimiser (too close to the existing tool geoptimiser). centreOptimiser? recentrer? onering (torulethemall)?

Ah, the name is a tad too obvious to pick I guess. I've changed it to centreOptimiser.

> The intro should say that this works with sharp diffraction rings, so people can quickly tell if they have suitable data to use it.

Clarified this a bit more in the tutorial notebook.

> I'll look at the avg_frame function soon, but I'd prefer to make the example work through from loading data and assembling a frame with EXtra-geom, even if this requires a bit of extra code in the notebook or makes the pictures slightly less neat. The handy function with AGIPD-specific adjustments seems like a distraction from the key message.

I removed the section in the tutorial notebook that calls the `avg_frame` function and basically just pasted the code in directly. I agree, it's better to have the code there so that it's obvious what's happening to the users.

> If the optimiser takes a geometry object, maybe it would be neater for it to take a 3D array and do the assembly as part of its setup.

That could work, but I need more than just the assembled data for the geometry. It requires `pixel_size` and `modules` as well. I could put those in as arguments to the function, at that point it'd be replacing one argument with three.

> missaligned -> misaligned

Corrected the misspelling.

> Is it possible to briefly describe what the optimiser does in the notebook? It's making the highest peak in the azimuthal integration as high as possible, right? Or is it more complex now?

Added a short explanation in the tutorial notebook.

> Corrected geometry: is this available from res - something like res.new_geom` - rather than copying and pasting it back into code?

I've added the corrected geometry in as an output in the result tuple, it also gets printed off (to make it easier to copy/paste, which you would want to do once you've found the corrected geometry).

Right now, due to the change to scipy's more accurate optimiser, the results have a lot of decimal places. Is it worth rounding this off to make it more readable, or should I just leave it as it is?

> "...which typically do not have to be looked at." - a bit tortured! "...which you typically don't need to look at"?

Section removed with change to scipy optimisers.

> In the second 'unrolled' image, is it worth drawing a vertical line to show how neatly the rings now unroll? It could be offset a few pixels so it doesn't obscure the lines, or use a dotted line. Maybe even draw it under the image so you see it in the gaps.

Good idea!

> Let's add a header (maybe "Troubleshooting") where the discussion of examining the optimisation process starts. It's definitely good to include this, but it's also nice to have a clear marker where the standard workflow ends.

Skopt has nice plots for this, but the scipy global optimisers don't, and they don't log much data for what happened during the optimisation, so with the move to these optimisers it seems a bit more difficult to explain the troubleshooting/diagnostic steps. For now I just removed these from the notebook.

RobertRosca commented on geoAssembler/optimiser/utility.py:
```diff
@@ -0,0 +1,140 @@
+
+        self.centre = [centre_geom[0], centre_geom[1]]
+
+        self.detector = Detector(
```
Hmm the `setpyfai` thing, for some reason, loses the detector/pixel info so I have to add it back in each time, but it is a bit cleaner this way. Cheers.

takluyver commented:
> > If the optimiser takes a geometry object, maybe it would be neater for it to take a 3D array and do the assembly as part of its setup.
>
> That could work, but I need more than just the assembled data for the geometry.

To be clear, I meant:

```python
# Instead of this:
optimiser = CentreOptimiser(geom, assembled_frame, sample_dist_m=0.2)

# How about this:
optimiser = CentreOptimiser(geom, modules_stack, sample_dist_m=0.2)
```

It's not a big difference, but it's slightly surprising that the caller has to assemble the image itself when it's passing the geometry to the optimiser anyway. Passing an un-assembled stack would also be consistent if we ever add optimisers for e.g. quadrant positions.

takluyver commented:
> The optimiser works by maximising the value of the 1D integration result. 

The 1D integration result is a vector of values. We're maximising the maximum, right? "...maximising the highest value in the 1D integration result" ?

takluyver commented on geoAssembler/optimiser/utility.py:
```diff
@@ -0,0 +1,136 @@
+
+def avg_frame(run: DataCollection, geom: DetectorGeometryBase,
+              train_index: int, pulse_pattern=None,
+              masking=True,) -> Tuple[np.ndarray, Tuple]:
```
This isn't now used or mentioned anywhere, I think? So maybe it should be dropped (or archived somewhere until we put the bits into other code). If we keep it here, I think the name should at least mention AGIPD1M.

takluyver commented on geoAssembler/tests/test_centre_optimiser.py:
```diff
@@ -0,0 +1,43 @@
+    ])
+
+frame_path = os.path.dirname(geoAssembler.__file__) + "/tests/optimiser-test-frame.npy"
+frame = np.load(frame_path)
```
Let's move the data loading into the test (or into a fixture, if you want to be fancy), so it doesn't happen on import.

Also, how is it 10 MB? 1 corrected AGIPD frame should be about 4 MB with no compression at all (4 bytes per pixel, 1 megapixel).

RobertRosca replied:
Moved the import into the function now.

It's 10 MB as I applied an edge mask to the frame, and apparently `np.full((dims...), np.nan)` returns a float64, which I didn't realise, so when it was applied the whole frame became a float64. I use BFG Repo-Cleaner to delete the file and save it as a float32 if you'd like.

takluyver replied:
The proc data is normally float32, so I think it's OK to save it like that. No need for rewriting history at this point, I think, so long as we remember to squash commits when merging the PR.

But can we consider using a stack of module data first - that would also change what we save.

takluyver commented on geoAssembler/tests/test_centre_optimiser.py:
```diff
@@ -0,0 +1,43 @@
+def test_integrator():
+    optimiser = centreOptimiser.CentreOptimiser(geom, frame, sample_dist_m=0.2)
+
+    misaligned_2dint = optimiser.integrate2d(frame).intensity
+    misaligned_2dint_r = optimiser.integrate2d(frame).radial
+    misaligned_2dint_a = optimiser.integrate2d(frame).azimuthal
```
Would it not be better to call `integrate2d()` once and then examine the different bits of the result?

takluyver commented on geoAssembler/tests/test_centre_optimiser.py:
```diff
@@ -0,0 +1,43 @@
+    misaligned_2dint_r = optimiser.integrate2d(frame).radial
+    misaligned_2dint_a = optimiser.integrate2d(frame).azimuthal
+
+    assert misaligned_2dint.shape == (957, 832)
+    assert misaligned_2dint_r.shape == (832,)
+    assert misaligned_2dint_a.shape == (957,)
```
Asserting on exact numbers like this seems like it's probably quite brittle - if some detail of the implementation changes, it could presumably change by a pixel or two.

takluyver commented on geoAssembler/tests/test_centre_optimiser.py:
```diff
@@ -0,0 +1,43 @@
+    misaligned_1dint = np.nanmean(misaligned_2dint, axis=0)[100:-100]
+    misaligned_1dint_x = optimiser.integrate2d(frame).radial[100:-100]
+
+    #  Kinda crappy test...
+    # assert hash(misaligned_1dint.data.tobytes()) == 993467963203778007
+
```
```suggestion
```

takluyver replied:
i.e. I agree with your comment ;-)

takluyver commented on geoAssembler/tests/test_centre_optimiser.py:
```diff
@@ -0,0 +1,43 @@
+        misaligned_1dint == np.max(misaligned_1dint
+    ))[0][0]
+
+    assert brightest_ring_idx == 106
```
Again, this seems very specific - maybe check that it's in a range like 95 - 105.

takluyver commented on setup.py:
```diff
@@ -60,7 +60,8 @@ def find_files(directory):
           'PyQt5-sip >= 12.7.0, <= 13.0.0',
-          'scipy'
+          'scipy',
+          'xarray <= 0.14.1'
```
```suggestion
          'xarray <= 0.14.1',
```

Trailing commas are nice, then the next addition only needs to change one line.

takluyver commented on setup.py:
```diff
@@ -60,7 +60,8 @@ def find_files(directory):
           'PyQt5-sip >= 12.7.0, <= 13.0.0',
-          'scipy'
+          'scipy',
+          'xarray <= 0.14.1',
```
Do you really mean `<=`? I.e. this won't work with xarray 0.15?

takluyver commented:
I'd still favour passing in a stack of module data rather than a pre-assembled image - see [this comment](https://github.com/European-XFEL/geoAssembler/pull/22#issuecomment-628624061) - unless there's a good reason not to.

Otherwise, I think this is in good shape.

RobertRosca commented:
> To be clear, I meant:
> 
> # Instead of this:
> optimiser = CentreOptimiser(geom, assembled_frame, sample_dist_m=0.2)
> 
> # How about this:
> optimiser = CentreOptimiser(geom, modules_stack, sample_dist_m=0.2)
> 
> It's not a big difference, but it's slightly surprising that the caller has to assemble the image itself when it's passing the geometry to the optimiser anyway. Passing an un-assembled stack would also be consistent if we ever add optimisers for e.g. quadrant positions.

Ah okay I get what you mean now. I let it use the assembled images as, from what I've seen, users prefer working with arrays of frames rather than arrays of modules, and conceptually it makes sense for the centre optimiser to just take in the assembled image instead of the modules.

I already wrote (but didn't commit) a quadrant assembler which does take in the module stack for the reasons you mentioned, so I'll change the centre optimiser over to taking the quadrants instead as well (after lunch :stuck_out_tongue:)

takluyver commented:
A 2D image is definitely a comfortable, familiar thing, but I think we should encourage users to handle detector data as stacks (3D arrays) as far through the analysis as possible - it's often more efficient and more precise to process the data like this and only assemble it when you want to view it.

(Obviously this gets into deeper questions - how far should we encourage what we see as good practices vs just doing what users want? There's a balance to be maintained, and I think I lean more towards 'good practices' than some others.)

takluyver commented:
Can you make the test image float32?

tmichela commented on geoAssembler/optimiser/centre.py:
```diff
@@ -0,0 +1,109 @@
+    """
+    def __init__(self, geom: DetectorGeometryBase,
+                 stacked_mean: np.ndarray, sample_dist_m: Union[int, float],
+                 unit: str = "2th_deg"):
```
Add maybe a docstring that describes what are the input arguments.

tmichela replied:
Also `stack_mean` is maybe misleading, as I suppose it does not have to be a mean image to be usable?

takluyver commented:
Thanks @RobertRosca for being patient with my code review for this.

RobertRosca commented:
Thanks for reviewing it! Next time I'll [cheat by remembering your comments](https://i.imgur.com/ZTH8tv4.png) and writing it so there are fewer things to fix 😂
